### PR TITLE
Add GitHub actions for issues

### PR DESCRIPTION
**Pull request by ec4n6**
## Proposed Changes

Add new actions to Github app: Create Issue Comment, Close Issue, Add Issue Label

## PR Requirements

Developers, verify you have completed the following items by checking them off:

### Style

Review the [style guide](https://komand.github.io/python/style.html)

- [x] For dependencies, pin [OS package](https://komand.github.io/python/style.html#dockerfile) and [Python package](https://komand.github.io/python/style.html#requirements-txt) versions
- [x] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [x] For size, use the [slim SDK images](https://komand.github.io/python/sdk.html#sdk-versions) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [x] For error handling, use of [PluginException](https://komand.github.io/python/error_handling.html#plugin-exceptions) and [ConnectionTestException](https://komand.github.io/python/error_handling.html#connection-exceptions)
- [x] For logging, use [self.logger](https://komand.github.io/python/sdk.html#logging)
- [x] For docs, use [changelog style](https://komand.github.io/python/style.html#changelog)
- [x] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://komand.github.io/python/style.html#tests) in the `tests/` directory created with `./run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `./run -T tests/example.json -d -j`
    * All actions/triggers shortcut `./run -T all -d -j` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `./run -R tests/example.json -d -j`
    * All actions/triggers shortcut `./run -R all -d -j` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `make validate` and make sure everything passes

<details>make validate

```bash
/insightconnect-plugins/github$ make validate
[*] Use ``make menu`` for available targets
[*] Including available Makefiles: ../tools/Makefiles/Colors.mk ../tools/Makefiles/Helpers.mk
--
[*] Running validators
[*] Validating plugin at .

[*] Running Integration Validators...
[*] Executing validator SpecPropertiesValidator
[*] Executing validator FilesValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator MakefileValidator
[*] Executing validator VersionValidator
[*] Executing validator HelpValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator DockerValidator
[*] Executing validator LoggingValidator
[*] Plugin successfully validated!

----
[*] Total time elapsed: 4059.3510000000006ms

[*] Validating spec with js-yaml
[SUCCESS] Passes js-yaml spec check


[*] Validating markdown...
[SUCCESS] Passes markdown linting

[*] Validating python files for style...
./komand_github/connection/connection.py:28:13: F841 local variable 'user_info' is assigned to but never used
./komand_github/triggers/issue/trigger.py:38:29: E231 missing whitespace after ','
./komand_github/triggers/issue/trigger.py:81:9: E265 block comment should start with '# '
./komand_github/triggers/issue/trigger.py:85:26: E261 at least two spaces before inline comment
./komand_github/triggers/issue/trigger.py:85:26: E262 inline comment should start with '# '
./komand_github/triggers/issue/trigger.py:86:13: E265 block comment should start with '# '
./komand_github/triggers/issue/trigger.py:87:13: E265 block comment should start with '# '
./komand_github/triggers/issue/trigger.py:88:13: E265 block comment should start with '# '
./komand_github/triggers/issue/trigger.py:90:15: E222 multiple spaces after operator
./komand_github/triggers/issue/trigger.py:96:12: E221 multiple spaces before operator
./komand_github/triggers/issue/trigger.py:97:13: E221 multiple spaces before operator
./komand_github/triggers/issue/trigger.py:118:17: E265 block comment should start with '# '
./komand_github/triggers/issue/trigger.py:125:17: E265 block comment should start with '# '
./komand_github/triggers/issue/trigger.py:126:17: E265 block comment should start with '# '
./komand_github/triggers/issue/trigger.py:134:29: E265 block comment should start with '# '
./komand_github/actions/get_my_issues/action.py:21:21: E201 whitespace after '{'
./komand_github/actions/get_my_issues/action.py:21:59: E202 whitespace before '}'
./komand_github/actions/create_issue_comment/action.py:2:1: F401 '.schema.Input' imported but unused
./komand_github/actions/create_issue_comment/action.py:2:1: F401 '.schema.Output' imported but unused
./komand_github/actions/create_issue_comment/action.py:25:39: W292 no newline at end of file
./komand_github/actions/close_issue/action.py:2:1: F401 '.schema.Input' imported but unused
./komand_github/actions/close_issue/action.py:2:1: F401 '.schema.Output' imported but unused
./komand_github/actions/close_issue/action.py:25:18: W292 no newline at end of file
./komand_github/actions/user/action.py:16:7: E111 indentation is not a multiple of four
./komand_github/actions/user/action.py:17:7: E111 indentation is not a multiple of four
./komand_github/actions/user/action.py:19:11: E111 indentation is not a multiple of four
./komand_github/actions/user/action.py:21:7: E111 indentation is not a multiple of four
./komand_github/actions/user/action.py:22:7: E111 indentation is not a multiple of four
./komand_github/actions/user/action.py:25:7: E111 indentation is not a multiple of four
./komand_github/actions/user/action.py:26:7: E111 indentation is not a multiple of four
./komand_github/actions/user/action.py:27:7: E111 indentation is not a multiple of four
./komand_github/actions/user/action.py:28:7: E111 indentation is not a multiple of four
./komand_github/actions/block_user/action.py:27:13: F841 local variable 'data' is assigned to but never used
./komand_github/actions/block_user/action.py:31:25: E201 whitespace after '{'
./komand_github/actions/block_user/action.py:31:41: E202 whitespace before '}'
./komand_github/actions/block_user/action.py:33:17: E201 whitespace after '{'
./komand_github/actions/block_user/action.py:33:34: E202 whitespace before '}'
./komand_github/actions/utils.py:3:1: E302 expected 2 blank lines, found 1
./komand_github/actions/add_issue_label/action.py:2:1: F401 '.schema.Input' imported but unused
./komand_github/actions/add_issue_label/action.py:2:1: F401 '.schema.Output' imported but unused
./komand_github/actions/add_issue_label/action.py:25:18: W292 no newline at end of file
./komand_github/actions/create/action.py:2:1: F401 'github' imported but unused
./komand_github/actions/create/action.py:7:1: E303 too many blank lines (3)
./komand_github/actions/create/action.py:8:3: E111 indentation is not a multiple of four
./komand_github/actions/create/action.py:9:7: E111 indentation is not a multiple of four
./komand_github/actions/create/action.py:16:3: E303 too many blank lines (2)
./komand_github/actions/create/action.py:16:3: E111 indentation is not a multiple of four
./komand_github/actions/create/action.py:18:7: E111 indentation is not a multiple of four
./komand_github/actions/create/action.py:19:7: E111 indentation is not a multiple of four
./komand_github/actions/create/action.py:21:7: E111 indentation is not a multiple of four
./komand_github/actions/create/action.py:22:7: E111 indentation is not a multiple of four
./komand_github/actions/get_issues_by_repo/action.py:23:21: E201 whitespace after '{'
./komand_github/actions/get_issues_by_repo/action.py:23:59: E202 whitespace before '}'
./komand_github/actions/unblock_user/action.py:27:13: F841 local variable 'data' is assigned to but never used
./komand_github/actions/unblock_user/action.py:31:25: E201 whitespace after '{'
./komand_github/actions/unblock_user/action.py:31:41: E202 whitespace before '}'
./komand_github/actions/unblock_user/action.py:33:17: E201 whitespace after '{'
./komand_github/actions/unblock_user/action.py:33:34: E202 whitespace before '}'
[FAIL] Fails flake8 linting

[*] Validating python files for security vulnerabilities...
[main]	INFO	profile include tests: None
[main]	INFO	profile exclude tests: None
[main]	INFO	cli include tests: None
[main]	INFO	cli exclude tests: None
[main]	INFO	running on Python 3.7.3
54 [0.. 50.. ]
Run started:2019-08-05 14:35:30.567617

Test results:
	No issues identified.

Code scanned:
	Total lines of code: 5907
	Total lines skipped (#nosec): 0

Run metrics:
	Total issues (by severity):
		Undefined: 0.0
		Low: 0.0
		Medium: 0.0
		High: 0.0
	Total issues (by confidence):
		Undefined: 0.0
		Low: 0.0
		Medium: 0.0
		High: 0.0
Files skipped (0):
[SUCCESS] Passes bandit security checks
```
</details>

2. Run the assessment tool: `./run -A -R all -T all`. For single action validation: `./run -A -R tests/my_action.json -T tests/my_action.json`
3. Copy (`./run ... | pbcopy`) and paste the output in **a new post** on this PR.
4. Add UI screenshot of the workflow used for testing
5. Add UI screenshot of the job output used for testing
6. Add UI screenshot of the artifact (See rules in UI Checklist) used for testing

### CODE REVIEW

File: github/Makefile [hunk 1]
```diff
@@ -45,9 +45,9 @@ regenerate:
 
 export: image
 	$(info [$(YELLOW)*$(NORMAL)] Exporting docker image)
-	@printf "\n ---> Exporting Docker image to ./$(VENDOR)_$(NAME)_$(VERSION).tar\n"
-	@docker save $(VENDOR)/$(NAME):$(VERSION) | gzip > $(VENDOR)_$(NAME)_$(VERSION).tar
+	@printf "\n ---> Exporting Docker image to ./$(NAME).img\n"
+	@docker save $(VENDOR)/$(NAME):$(VERSION) | gzip > $(NAME)_$(VERSION).img
```
Comment (mrinehart-r7) [inline github/Makefile]: This Makefile line is out of date, can you run `update-tools.sh` in the `tools/` directory, then regenerate the plugin?

File: github/help.md [hunk 2]
```diff
@@ -975,6 +1027,7 @@ Examples:
 * 1.2.1 - Update connection documentation and add example outputs to remaining actions and triggers
 * 2.0.0 - Rename "User" action to "Get User"
 * 2.0.1 - Fix missing Search action | Pin pygithub and python-dateutil libraries | Update to use the `komand/python-3-37-slim-plugin` Docker image to reduce plugin size | Enable verification of SSL/TLS certificates for github.com
+* 2.0.2 - New actions: Create Issue Comment, Close Issue, Add Issue Label
```
Comment (mrinehart-r7) [inline github/help.md]: 2.1.0, minor bump due to backwards compatible additions

File: github/komand_github/actions/add_issue_label/action.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+    def __init__(self):
+        super(self.__class__, self).__init__(
+                name='add_issue_label',
+                description=Component.DESCRIPTION,
+                input=AddIssueLabelInput(),
+                output=AddIssueLabelOutput())
+
+    def run(self, params={}):
+        if params.get('organization') and params.get('repository'):
```
Comment (mrinehart-r7) [inline github/komand_github/actions/add_issue_label/action.py]: Whenever you access `params` dict use the I/O constants that are codegenerated, ie `Input.ORGANIZATION`

File: github/plugin.spec.yaml [hunk 4]
```diff
@@ -1142,6 +1142,70 @@ actions:
         description: URL
         type: string
         required: false
+  create_issue_comment:
+    title: Create Issue Comment
+    description: Create a issue comment
```
Comment (mrinehart-r7) [inline github/plugin.spec.yaml]: `Create an issue comment`

File: github/plugin.spec.yaml [hunk 5]
```diff
@@ -1142,6 +1142,70 @@ actions:
...
+        required: false
+    output:
+      url:
+        title: URL
+        description: URL
+        type: string
+        required: false
+  close_issue:
+    title: Close Issue
+    description: Close Issue
```
Comment (mrinehart-r7) [inline github/plugin.spec.yaml]: `Close issue`

File: github/plugin.spec.yaml [hunk 6]
```diff
@@ -1142,6 +1142,70 @@ actions:
...
+        description: Repository to post issue
+        required: true
+      organization:
+        type: string
+        description: Organizational owner of repository
+        required: false
+    output:
+  add_issue_label:
+    title: Add Issue Label
+    description: Adds Label to an Issue
```
Comment (mrinehart-r7) [inline github/plugin.spec.yaml]: `Adds a label to an issue`

File: github/komand_github/actions/add_issue_label/action.py (hunk 3 above)
Comment (jschipp-r7) [reply github/komand_github/actions/add_issue_label/action.py]: Here's more information. https://komand.github.io/python/sdk.html#actions

File: github/plugin.spec.yaml [hunk 7]
```diff
@@ -1142,6 +1142,70 @@ actions:
         description: URL
         type: string
         required: false
+  create_issue_comment:
+    title: Create Issue Comment
+    description: Create an issue commentt
```
Comment (jschipp-r7) [inline github/plugin.spec.yaml]: Typo
```suggestion
    description: Create an issue comment
```
And then run `make regenerate` so the changes reflect in the schema files

File: github/komand_github/actions/add_issue_label/action.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+            g = self.connection.github_user
+            issue = g.get_organization(params.get('organization')).get_repo(params.get('repository')).get_issue(int(params.get('issue_number')))
+        else:
+            g = self.connection.user
+            issue = g.get_repo(params.get('repository')).get_issue(int(params.get('issue_number')))
+
+        issue_params = params.get("label")
+        issue = issue.add_to_labels(issue_params)
+        return {}
```
Comment (jschipp-r7) [inline github/komand_github/actions/add_issue_label/action.py]: We really should return an object, a status, or similar so workflow decisions can be mode off the output.

Same for other actions that are missing output. Needs a plugin spec output schema, updated documentation table for output and example JSON output in `help.md`.

Comment (ec4n6) [reply github/komand_github/actions/add_issue_label/action.py]: @jschipp-r7 The function `[add_to_labels](https://pygithub.readthedocs.io/en/latest/github_objects/Issue.html#github.Issue.Issue.add_to_labels)` returns `None`.

Should I add a try/expect?

Comment (jschipp-r7) [reply github/komand_github/actions/add_issue_label/action.py]: Oh, sorry, I didn't realize that. You're on the right path then. What we could do is have an output item called `success` or `status` and it returns a boolean. False if there's an exception when the method is called `{ 'return': False }`, True if all good.

File: github/komand_github/actions/add_issue_label/action.py (hunk 3 above)
Comment (ec4n6) [reply github/komand_github/actions/add_issue_label/action.py]: Fixed. Appreciate the context :)

**Comment by ec4n6** [comment]
@mrinehart-r7 @jschipp-r7 All the changes should be in place. Please take another look at your connivence. Thanks for the feedback and prompt responses

**Pull request merged**